### PR TITLE
Add a new goodput elastic utils function record_slice_state()

### PR DESCRIPTION
**Pull request by YixuanWang-99**
Update MaxText's [utils/elastic_utils.py](https://github.com/AI-Hypercomputer/maxtext/blob/main/src/maxtext/utils/elastic_utils.py) to add a new record_slice_state helper that queries live slice counts and logs them to the GoodputRecorder. This helper maps the core metrics directly from Pathways and JAX states. 

# Tests

Passed synthetic tests in tests/unit/elastic_utils_test.py.

Screenshot 2026-06-09 at 9 05 00 PM

Test with a small synthetic workflow and the [logs](https://pantheon.corp.google.com/logs/query;query=logName%3D%22projects%2Fcloud-tpu-multipod-dev%2Flogs%2Fgoodput_ELASTIC_TEST8%22;cursorTimestamp=2026-06-29T17:59:06.101871Z;duration=PT1H?e=13803378&mods=allow_workbench_image_override&project=cloud-tpu-multipod-dev) are good:

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/utils/elastic_utils.py [hunk 1]
```diff
@@ -28,13 +28,30 @@
 pending_elastic_event_type = None
 
 
+def record_slice_state(recorder, active_slices_override: int | None = None) -> None:
+  """Queries live slice counts and logs them to the GoodputRecorder."""
```
Comment (dipannita08) [inline src/maxtext/utils/elastic_utils.py]: Did you add a patch to instantiate ElasticGoodputRecorder? Essentially we would either need to add the API to the base class or instantiate a ElasticGoodputRecorder explicitly for elastic workload inside maxtext/common/goodput.py based on configs to avoid the AttributeError.

File: src/maxtext/utils/elastic_utils.py [hunk 2]
```diff
@@ -28,13 +28,30 @@
...
+  if recorder is None or not pathwaysutils.is_pathways_backend_used() or elastic_manager is None:
+    return
+
+  available_slices = len(pathwaysutils.elastic.get_active_slice_indices())
+  active_slices = active_slices_override if active_slices_override is not None else len(elastic_manager.active_slice_indices)
+  total_slices = len(elastic_manager.slice_to_devices)
+
+  if recorder:
```
Comment (dipannita08) [inline src/maxtext/utils/elastic_utils.py]: recorder should be non-None here because of your previous check in Ln 33.

File: src/maxtext/utils/elastic_utils.py [hunk 3]
```diff
@@ -46,6 +62,8 @@ def record_elastic_wait_end_and_reinit_start(recorder) -> None:
   pending_elastic_event_type = None
   if recorder:
     recorder.record_custom_badput_event_end_time(custom_badput_event_type=event_type)
+    recorder.record_custom_badput_event_start_time(custom_badput_event_type='elastic_reinitialization')
```
Comment (dipannita08) [inline src/maxtext/utils/elastic_utils.py]: Did we add this extra start_time recorder call for elastic_reinit by accident? Looks like Ln67 has another one.

File: src/maxtext/utils/elastic_utils.py [hunk 4]
```diff
@@ -28,13 +28,29 @@
...
+      total_slices=total_slices
+  )
+
 def record_elastic_event_start(recorder, config) -> None:
   """Records start of an elastic scale up event."""
   global pending_elastic_event_type
   event_type = "elastic_scale_up" if is_scale_up_event(config) else "elastic_slice_down"
   pending_elastic_event_type = event_type
   if recorder:
     recorder.record_custom_badput_event_start_time(custom_badput_event_type=event_type)
```
Comment (dipannita08) [inline src/maxtext/utils/elastic_utils.py]: [name update]: Update to the ElasticRecorder version: `recorder.record_elastic_wait_start_time(event_type=event_type)`

File: src/maxtext/utils/elastic_utils.py [hunk 5]
```diff
@@ -46,14 +62,19 @@ def record_elastic_wait_end_and_reinit_start(recorder) -> None:
...
   pending_reinit_recorder = recorder
 
 
 def record_elastic_reinit_end() -> None:
   """Records end of elastic reinitialization event."""
   global pending_reinit_recorder
   if pending_reinit_recorder is not None:
+    pending_reinit_recorder.record_custom_badput_event_end_time(
+        custom_badput_event_type='elastic_reinitialization'
+    )
```
Comment (dipannita08) [inline src/maxtext/utils/elastic_utils.py]: [name change]: update to `pending_reinit_recorder.record_elastic_reinit_end_time()`

File: src/maxtext/utils/elastic_utils.py [hunk 6]
```diff
@@ -46,14 +62,19 @@ def record_elastic_wait_end_and_reinit_start(recorder) -> None:
...
+    record_slice_state(pending_reinit_recorder)
     pending_reinit_recorder.record_custom_badput_event_end_time(custom_badput_event_type="elastic_reinitialization")
```
Comment (dipannita08) [inline src/maxtext/utils/elastic_utils.py]: Looks like another duplicate elastic_reinit_end_time call - we can remove it.

File: src/maxtext/utils/elastic_utils.py [hunk 7]
```diff
@@ -46,14 +62,19 @@ def record_elastic_wait_end_and_reinit_start(recorder) -> None:
   pending_elastic_event_type = None
   if recorder:
     recorder.record_custom_badput_event_end_time(custom_badput_event_type=event_type)
```
Comment (dipannita08) [inline src/maxtext/utils/elastic_utils.py]: [name change]: recorder.record_elastic_wait_end_time(event_type=event_type)

File: src/maxtext/utils/elastic_utils.py [hunk 8]
```diff
@@ -46,14 +62,19 @@ def record_elastic_wait_end_and_reinit_start(recorder) -> None:
...
-    recorder.record_custom_badput_event_start_time(custom_badput_event_type="elastic_reinitialization")
+    recorder.record_custom_badput_event_start_time(custom_badput_event_type='elastic_reinitialization')
```
Comment (dipannita08) [inline src/maxtext/utils/elastic_utils.py]: [name change]: `recorder.record_elastic_reinit_start_time()`

File: src/maxtext/utils/elastic_utils.py [hunk 9]
```diff
@@ -28,13 +28,29 @@
 pending_elastic_event_type = None
 
 
+def record_slice_state(recorder, active_slices_override: int | None = None) -> None:
+  """Queries live slice counts and logs them to the GoodputRecorder."""
```
Comment (dipannita08) [inline src/maxtext/utils/elastic_utils.py]: [nit] Records (instead of Queries)?

File: src/maxtext/utils/elastic_utils.py [hunk 10]
```diff
@@ -28,13 +28,29 @@
...
+  if recorder is None or not pathwaysutils.is_pathways_backend_used() or elastic_manager is None:
```
Comment (dipannita08) [inline src/maxtext/utils/elastic_utils.py]: Let's add a hassattr guard here incase workload is configured incorrectly we can prevent crashes. Add a 'or not hasattr(recorder, 'record_elastic_slice_counts')' to this to bail early?

**Review by dipannita08 — COMMENTED** [review]
Thanks! This looks good, we also need to update [train.py](https://github.com/AI-Hypercomputer/maxtext/blob/c7b9871b8495418880e68be58bdcf0b857a47c3b/src/maxtext/trainers/pre_train/train.py#L787) to initialize available, active and total slice counts at the beginning of a run.  cc @lukebaumann

File: src/maxtext/utils/elastic_utils.py [hunk 11]
```diff
@@ -28,13 +28,29 @@
...
 
 
+def record_slice_state(recorder, active_slices_override: int | None = None) -> None:
+  """Records live slice counts and logs them to the GoodputRecorder."""
+  if recorder is None or not hasattr(recorder, 'record_elastic_slice_counts') or not pathwaysutils.is_pathways_backend_used() or elastic_manager is None:
+    return
+
+  available_slices = len(pathwaysutils.elastic.get_active_slice_indices())
+  active_slices = active_slices_override if active_slices_override is not None else len(elastic_manager.active_slice_indices)
+  total_slices = len(elastic_manager.slice_to_devices)
```
Comment (dipannita08) [inline src/maxtext/utils/elastic_utils.py]: Should this directly use `get_slice_to_devices` instead of the elastic_manager's cached value? 

`total_slices = len(pathwaysutils.elastic.get_slice_to_devices(jax.devices()))`

@lukebaumann what do you think?

**Comment by dipannita08** [comment]
> Thanks! This looks good, we also need to update [train.py](https://github.com/AI-Hypercomputer/maxtext/blob/c7b9871b8495418880e68be58bdcf0b857a47c3b/src/maxtext/trainers/pre_train/train.py#L787) to initialize available, active and total slice counts at the beginning of a run. cc @lukebaumann

Looks like we are initializing islice counts nside `create_goodput_recorder`, that sounds good to me.

**Review by dipannita08 — APPROVED** [review]

File: src/maxtext/common/goodput.py [hunk 12]
```diff
@@ -23,6 +23,7 @@
 import jax
 from enum import Enum
 from maxtext.utils import max_logging
+from maxtext.utils import pathwaysutils
```
Comment (dipannita08) [inline src/maxtext/common/goodput.py]: Shouldn't this be `import pathwaysutils` to pull from PyPi (latest seems to be pulled by tpu-requirements)?

File: tests/unit/elastic_utils_test.py [hunk 13]
```diff
@@ -327,12 +362,15 @@ def test_record_elastic_event_start(self):
     fake_recorder.record_custom_badput_event_start_time.assert_called_once_with(
```
Comment (dipannita08) [inline tests/unit/elastic_utils_test.py]: These should now fail since we updated the names. Can you update the names in the tests too?

File: src/maxtext/common/goodput.py [hunk 14]
```diff
@@ -123,6 +136,20 @@ def create_goodput_recorder(config):
     return None
   if config.enable_goodput_recording:
     logger_name = f"goodput_{config.run_name}"
-    recorder = goodput.GoodputRecorder(config.run_name, logger_name, jax.process_index() == 0)
+    
+    # Detect if we should use the elastic-aware recorder
+    use_elastic = getattr(config, 'elastic_enabled', False) and pathwaysutils.is_pathways_backend_used()
```
Comment (dipannita08) [inline src/maxtext/common/goodput.py]: [nit] Can we use the already defined `elastic_enabled(config)` helper for these?

File: tests/unit/elastic_utils_test.py [hunk 15]
```diff
@@ -327,12 +362,15 @@ def test_record_elastic_event_start(self):
...
         custom_badput_event_type="elastic_slice_down"
     )
+    fake_recorder.record_elastic_slice_counts.assert_called_once()
+    self.assertEqual(elastic_utils.pending_elastic_event_type, 'elastic_slice_down')
     self.assertEqual(elastic_utils.pending_elastic_event_type, "elastic_slice_down")
```
Comment (dipannita08) [inline tests/unit/elastic_utils_test.py:359]: [nit] Looks like we are accidentally asserting this twice.

File: src/maxtext/common/goodput.py [hunk 16]
```diff
@@ -66,18 +67,30 @@ def maybe_monitor_goodput(config):
         enable_gcp_goodput_metrics=config.enable_gcp_goodput_metrics,
         enable_gcp_step_deviation_metrics=config.enable_gcp_step_deviation_metrics,
     )
-    goodput_monitor = monitoring.GoodputMonitor(
+    monitor_class = monitoring.GoodputMonitor
+    use_elastic = getattr(config, 'elastic_enabled', False) and pathwaysutils.is_pathways_backend_used()
```
Comment (lukebaumann) [inline src/maxtext/common/goodput.py]: Use the config value directly instead of getattr. If we do not validate the is_pathways_backend_used already when elastic_enabled and enable_single_controller, we need to make that change.

72 would then become
if config.elastic_enabled: 

and no need for 71.

File: src/maxtext/common/goodput.py [hunk 17]
```diff
@@ -66,18 +67,30 @@ def maybe_monitor_goodput(config):
...
+    if use_elastic:
+      try:
+        from ml_goodput_measurement import monitoring_elastic
+        monitor_class = monitoring_elastic.ElasticGoodputMonitor
+      except ImportError:
```
Comment (lukebaumann) [inline src/maxtext/common/goodput.py:76]: I think a log would be good here

File: src/maxtext/common/goodput.py [hunk 18]
```diff
@@ -123,6 +137,19 @@ def create_goodput_recorder(config):
     return None
   if config.enable_goodput_recording:
     logger_name = f"goodput_{config.run_name}"
-    recorder = goodput.GoodputRecorder(config.run_name, logger_name, jax.process_index() == 0)
+    
+    # Detect if we should use the elastic-aware recorder
+    recorder = None
```
Comment (lukebaumann) [inline src/maxtext/common/goodput.py]: consider simplifying:

```
if not config.enable_goodput_recording:
  return None

logger_name = ...

if config.elastic_enabled:
  try:
    ...
    recorder = goodput_elastic.ElasticGoodputRecorder(...)
    ...
  except ImportError:
    max_logging.log("Could not create elastic goodput recorder" ...)
  else:
    return recorder

return goodput.GoodputRecorder(...)
```

**Review by NuojCheng — APPROVED** [review]
